onnxruntime/core/providers/coreml/builders/impl/binary_op_builder.cc

```diff
@@ -19,24 +19,58 @@ class BinaryOpBuilder : public BaseOpBuilder {
  private:
 #ifdef __APPLE__
   [[nodiscard]] Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
-                               const logging::Logger& logger) const override;
+                                             const logging::Logger& logger) const override;
 #endif
   // Operator support related
   int GetMinSupportedOpSet(const Node& node) const override;
+
+  bool HasSupportedInputsImpl(const Node& node, const logging::Logger& logger) const override;
 };
 
+static bool CheckIfBothInputShapesMatch(const Node& node, const logging::Logger& logger) {
+  const auto& input_defs = node.InputDefs();
+  std::vector<int64_t> input_shape1;
+  if (!GetShape(*input_defs[0], input_shape1, logger))
+    return false;
+
+  std::vector<int64_t> input_shape2;
+  if (!GetShape(*input_defs[1], input_shape2, logger))
+    return false;
+
+  return (input_shape1.size() == input_shape2.size() &&
+          std::equal(input_shape1.begin(), input_shape1.end(), input_shape2.begin()));
```

Comment:
nit: FYI std::vector has [operator==](https://en.cppreference.com/w/cpp/container/vector/operator_cmp)
```suggestion
  return input_shape1 == input_shape2;
```